src/main/java/org/zalando/nakadi/service/timeline/TimelineSyncImpl.java

```diff
@@ -0,0 +1,365 @@
+package org.zalando.nakadi.service.timeline;
+
+import java.io.Closeable;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.apache.curator.framework.recipes.locks.InterProcessLock;
+import org.apache.curator.framework.recipes.locks.InterProcessSemaphoreMutex;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.Watcher;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Profile;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+import org.zalando.nakadi.repository.zookeeper.ZooKeeperHolder;
+import org.zalando.nakadi.util.UUIDGenerator;
+
+@Service
+@Profile("!test")
+public class TimelineSyncImpl implements TimelineSync {
+    private static class DelayedChange {
+        private final int version;
+        private final Set<String> lockedEventTypes;
+
+        private DelayedChange(final int version, final Collection<String> lockedEventTypes) {
+            this.version = version;
+            this.lockedEventTypes = new HashSet<>(lockedEventTypes);
+        }
+
+        @Override
+        public String toString() {
+            return "{version=" + version + ", lockedEventTypes=" + lockedEventTypes + '}';
+        }
+    }
+
+    private static final String ROOT_PATH = "/timelines";
+    private static final Charset CHARSET = Charset.forName("UTF-8");
+
+    private final ZooKeeperHolder zooKeeperHolder;
+    private final String thisId;
+    private final Set<String> lockedEventTypes = new HashSet<>();
+    private final Map<String, Integer> eventsBeingPublished = new HashMap<>();
+    private final Object localLock = new Object();
+    private final HashMap<String, List<Consumer<String>>> consumerListeners = new HashMap<>();
+    private final BlockingQueue<DelayedChange> queuedChanges = new LinkedBlockingQueue<>();
+    private final Logger log;
+
+    @Autowired
+    public TimelineSyncImpl(final ZooKeeperHolder zooKeeperHolder, final UUIDGenerator uuidGenerator) {
+        this.thisId = uuidGenerator.randomUUID().toString();
+        this.log = LoggerFactory.getLogger("timelines.sync." + thisId);
```

Comment:
I'm not against having id, but I would like to have an easy way to match it to actual instance. I suggest at least to log id right after generation so that in the logs we will be able to find which nakadi instance has that it.